public/app/features/scenes/dashboard/DashboardsLoader.test.ts

```diff
@@ -0,0 +1,239 @@
+import { VariableType } from '@grafana/schema';
+
+import { CustomVariable } from '../variables/variants/CustomVariable';
+import { DataSourceVariable } from '../variables/variants/DataSourceVariable';
+import { QueryVariable } from '../variables/variants/query/QueryVariable';
+
+import { createVariableFromLegacyModel } from './DashboardsLoader';
+
+describe('DashboardLoader', () => {
+  describe('variables migration', () => {
+    it('should migrate custom variable', () => {
+      const variable = {
+        current: {
+          selected: false,
+          text: 'a',
+          value: 'a',
+        },
+        hide: 0,
+        includeAll: false,
+        multi: false,
+        name: 'query0',
+        options: [
+          {
+            selected: true,
+            text: 'a',
+            value: 'a',
+          },
+          {
+            selected: false,
+            text: 'b',
+            value: 'b',
+          },
+          {
+            selected: false,
+            text: 'c',
+            value: 'c',
+          },
+          {
+            selected: false,
+            text: 'd',
+            value: 'd',
+          },
+        ],
+        query: 'a,b,c,d',
+        skipUrlSync: false,
+        type: 'custom' as VariableType,
+        rootStateKey: 'N4XLmH5Vz',
+        id: 'query0',
+        global: false,
+        index: 0,
+        state: 'Done',
+        error: null,
+        description: null,
+        allValue: null,
+      };
+
+      const migrated = createVariableFromLegacyModel(variable);
```

Comment:
If we're testing here, `DashboardLoader`. Shouldn't we test the migration through it? 

As I understand how we're doing our unit tests, the unit/subject of the test is the exported unit from the module. In the case of a scene object, we test the scene object and its integration with the scene. In this case, if we're testing the loader, I'd expect to initialize the loader with a template using the legacy model.

I also understand the DashboardLoader as it is, is hard to test. So I'm happy with this as soon as this serves as a signal for the DashboardLoader to be refactored in a way we can test this easily. WDYT?

Reply:
Yes, DashboardsLoader needs a refactor, then we can add a more verbose, integration test.